### DISCUSSION TITLE
problem running wbt_time_in_daylight()

### DISCUSSION THREAD

**Discussion by wiesehahn** [opengeos/whiteboxR#74] [Q&A] [answered]
not sure if this is an issue or if I am doing something wrong. 

I tried to run the `wbt_time_in_daylight()` function but get an  error.

Running `wbt_time_in_daylight(dem = dtm_path, output = here("wbt_time_in_daylight.tif"), lat = latitude, long = longitude)`  results in
```
Error running WhiteboxTools (time_in_daylight)
  whitebox.exe_path: "SomePath/R-4.1/x86_64-w64-mingw32/whitebox/WBT/whitebox_tools.exe"; File exists? TRUE
  Arguments: --run=time_in_daylight  --dem="SomePath/dtm.tif" --output="SomePath/wbt_time_in_daylight.tif" --lat=51.4 --long=9.8 --az_fraction=10 --max_dist=100 --utc_offset=0000 --start_day=1 --end_day=365 --start_time=000000 --end_time=235959 -v

System command had status 101
time_in_daylight - Elapsed Time: NA [did not run]
```
Whats the problem?

(btw in the documentation for `end_time` there is a mistake)

**Comment by Afrancioni** [comment]
Hello, 

Have you tried running the tool using a ":" for the start and end times parameters between the hours, minutes ,and seconds i.e. (00:00:00 and 23:59:59)? The tool documentation does use this format for start and end time. Alternatively you could use the values "sunrise" and "sunset" for the start time and end time to capture the full day. 

https://www.whiteboxgeo.com/manual/wbt_book/available_tools/geomorphometric_analysis.html#TimeInDaylight

**Reply by wiesehahn** [reply]
thanks, that did the trick

**Comment by brownag** [comment, marked as answer]
It seems that the colons are stripped out of the default values in the R function definitionwhen converted from whitebox_tools.py. So there is an issue here, thanks for bringing it to our attention. It looks like the documentation is correct, but the default arguments are not functional and haven't been for some time.

For instance, this works:
``` r
library(whitebox)
library(terra)

wbt_verbose(TRUE)
wbt_time_in_daylight(sample_dem_data(), output = "test.tif",  44.11856, -72.80839, max_dist=500, end_time='23:59:59')
#> time_in_daylight - Elapsed Time (excluding I/O): 0.566s
plot(rast("test.tif"))
```

![](https://i.imgur.com/gRBVRF8.png)

**Reply by wiesehahn** [reply]
good!
documentation is wrong in the sense that `end_time` is decribed as "starting hour to track shadows"

**Reply by brownag** [reply]
Ah good catch! The argument descriptions comes from [here](https://github.com/jblindsay/whitebox-tools/blob/31a909eae43e31ce8d6ffca6ae45ad2b0d30e99f/whitebox_tools.py#L4276), so that needs to be fixed on that repo. It will update in the R package when we build from the next version of WhiteboxTools

**Comment by jblindsay** [comment]
Good catch indeed. It's fixed in the user manual now.

**Comment by wiesehahn** [comment]
Sorry, got a new problem. 
Running `wbt_time_in_daylight(dem = dtm_path, output = "wbt_time_in_daylight.tif", lat = latitude, long = longitude, start_time = "sunrise", end_time = "sunset")` on my data finishes the job now without error messages, but the output contains nodata only. 
I thought it might be related to the projection since I had it in projected crs, but projecting it to 4326 did not change anything. 

Running your example works out though. 

So I thought I would give it a try on terra sample data 
```
library(terra)
library(sf)

f <- system.file("ex/elev.tif", package="terra")
r <- rast(f)

bbox<-st_as_sfc(st_bbox(r), crs = st_crs(r)) 
center<-st_centroid(bbox)
longitude <- round(center[[1]][1], 1)
latitude <- round(center[[1]][2], 1)

wbt_time_in_daylight(dem = r, output = here::here(),  lat = latitude, long = longitude, start_time = "sunrise", end_time = "sunset")
```

but this results again in 
```
Error running WhiteboxTools (time_in_daylight)
  whitebox.exe_path: "SomePath/R-4.1/x86_64-w64-mingw32/whitebox/WBT/whitebox_tools.exe"; File exists? TRUE
  Arguments: --run=time_in_daylight  --dem="SomePath/R/R-4.1.2/library/terra/ex/elev.tif" --output="SomePath/wbt_time_in_daylight.tif" --lat=49.8 --long=6.1 --az_fraction=10 --max_dist=100 --utc_offset=0000 --start_day=1 --end_day=365 --start_time=sunrise --end_time=sunset -v

System command had status 101
time_in_daylight - Elapsed Time: NA [did not run]
```

**Reply by brownag** [reply]
To get my example to run I had to increase the default value of `max_dist` from 100 to 500 (or at least 5*90). 

How are you running this R code? Are you not getting standard error output when run interactively something to the effect of "thread main panicked..."? To get that to show up in {reprex} output use `reprex::reprex(std_out_err = TRUE)`

For instance:
``` r
library(whitebox)
library(terra)
#> terra 1.5.21

wbt_verbose(TRUE)
wbt_time_in_daylight(sample_dem_data(), output = "test.tif",  44.11856, -72.80839, end_time='23:59:59')
#> 
#> Error running WhiteboxTools (time_in_daylight)
#>   whitebox.exe_path: "C:\Program Files\UniversityofGeulphWhiteboxTools1.4.0_WKS\WBT\whitebox_tools.exe"; File exists? TRUE
#>   Arguments: --run=time_in_daylight  --dem="C:/Users/Andrew.G.Brown/Documents/R/win-library/4.1/whitebox/extdata/DEM.tif" --output="test.tif" --lat=44.11856 --long=-72.80839 --az_fraction=10 --max_dist=100 --utc_offset=0000 --start_day=1 --end_day=365 --start_time=000000 --end_time=23:59:59 -v
#> System command had status 101
#> time_in_daylight - Elapsed Time: NA [did not run]
```
<summary>
Standard output and standard error
</summary>

``` sh
thread 'main' panicked at 'The maximum search distance parameter (--max_dist) must be larger than 5 x cell size.', src\tools\terrain_analysis\time_in_daylight.rs:512:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Reply by brownag** [reply]
I tested this on Windows; I see you may not be getting the standard error when run interactively. That is an issue. May need to consider moving `wbt_system_call()` over to {processx} instead of using `system()`

**Reply by wiesehahn** [reply]
in the R console I did not get the error. 
with {reprex} I get get it in your example
``` r
whitebox::wbt_time_in_daylight(whitebox::sample_dem_data(), output = "test.tif",  44.11856, -72.80839, end_time='23:59:59')
#> 
#> Error running WhiteboxTools (time_in_daylight)
#>   whitebox.exe_path: "C:/Users/jwiesehahn/arbeit/sideprojects/terrain_visualization/renv/library/R-4.1/x86_64-w64-mingw32/whitebox/WBT/whitebox_tools.exe"; File exists? TRUE
#>   Arguments: --run=time_in_daylight  --dem="C:/Users/jwiesehahn/arbeit/sideprojects/terrain_visualization/renv/library/R-4.1/x86_64-w64-mingw32/whitebox/extdata/DEM.tif" --output="test.tif" --lat=44.11856 --long=-72.80839 --az_fraction=10 --max_dist=100 --utc_offset=0000 --start_day=1 --end_day=365 --start_time=000000 --end_time=23:59:59 -v
#> System command had status 101
```

<sup>Created on 2022-04-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Standard output and standard error
</summary>

``` sh
thread 'main' panicked at 'The maximum search distance parameter (--max_dist) must be larger than 5 x cell size.', whitebox-tools-app\src\tools\terrain_analysis\time_in_daylight.rs:544:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

</details>


but not for my dataset (resolution is quite high and default value seems to be 100), so I guess this is not related to the problem of NoData values, right?

**Comment by wiesehahn** [comment]
I just tried to run `wbt_time_in_daylight()` again, which ended in an error. I remembered that I had an issue last time (2 years ago) as well. 
I am not entirely sure, but I think this time its again another error, nothing to do with colons and nothing to do with search radius. 

``` r
library(whitebox)
library(terra)
#> terra 1.7.71

wbt_verbose(TRUE)
wbt_time_in_daylight(sample_dem_data(), output = "test.tif",  44.11856, -72.80839, max_dist=500, end_time='23:59:59')
#> 
#> Error running WhiteboxTools (TimeInDaylight)
#>   whitebox.exe_path: "C:\Users\jwiesehahn\AppData\Roaming/R/data/R/whitebox/WBT/whitebox_tools.exe"; File exists? TRUE
#>   Arguments: --run=TimeInDaylight  --dem="C:/Program Files/R/R-4.3.3/library/whitebox/extdata/DEM.tif" --output="test.tif" --lat="44.11856" --long="-72.80839" --az_fraction=10 --max_dist=500 --utc_offset=00:00 --start_day=1 --end_day=365 --start_time=00:00:00 --end_time=23:59:59 -v
#> System command had status 101
#> *****************************
#> * Welcome to TimeInDaylight *
#> * Powered by WhiteboxTools  *
#> * www.whiteboxgeo.com       *
#> *****************************
#> Calculating the local almanac...
#> time_in_daylight - Elapsed Time: NA [did not run]
```

<sup>Created on 2024-04-25 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Standard output and standard error
</summary>

``` sh
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: ComponentRange { name: "ordinal", minimum: 1, maximum: 365, value: 366, conditional_range: true }', whitebox-tools-app\src\tools\terrain_analysis\time_in_daylight.rs:935:30
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

</details>

Any ideas?

**Reply by brownag** [reply]
I can confirm I get the same error on Linux. Via the R interface or via the command line. I tinkered with some of the parameters and same result.

I notice there were some changes (NaiveTime -> Time, NaiveDate -> Date) in how time is handled in this tool for WhiteboxTools v2.3: https://github.com/jblindsay/whitebox-tools/commit/4208336a1de591bf868ed26ec8d2bf1b552c9955#diff-4c6d03f9fd598a5f68bf881cd140eedc9642af3e07fa7bdc0d062f291adba587

Based on the error output `thread 'main' panicked at 'called Result::unwrap() on an Err value: ComponentRange { name: "ordinal", minimum: 1, maximum: 365, value: 366, conditional_range: true }', whitebox-tools-app\src\tools\terrain_analysis\time_in_daylight.rs:935:30` I would guess that this is an issue related to leap years. 

2024 is a leap year, 2023 is not. Previous code used 2020. It looks like the code (Line 934 in time_in_daylight.rs) might be hard-coded to use 2023, but is iterating over day of year 1 to 366. I would suggest reporting this on the main whitebox-tools repo: https://github.com/jblindsay/whitebox-tools/

https://github.com/jblindsay/whitebox-tools/blob/4208336a1de591bf868ed26ec8d2bf1b552c9955/whitebox-tools-app/src/tools/terrain_analysis/time_in_daylight.rs#L934-L941
